heracles/io.py

```diff
@@ -265,39 +265,48 @@ def _read_complex(hdu):
     return arr
 
 
+def _prepare_result_array(arr, order, size):
+    """Prepare result array for writing."""
+
+    if len(order) == 1:
+        return arr[0]
+    return np.transpose([np.pad(arr[i], (0, size - arr[i].size)) for i in order])
+
+
 def _write_result(fits, ext, key, result):
     """
     Write a result array to FITS.
     """
 
-    # keep ndarray subclasses or we would lose all Result attributes
-    result = np.asanyarray(result)
+    from heracles.result import normalize_result_axis, get_result_array
+
+    # original unsorted ell and axis
+    _ell = getattr(result, "ell", None)
+    _axis = normalize_result_axis(getattr(result, "axis", None), result, _ell)
 
-    # get ell axis
-    axis = getattr(result, "axis", result.ndim - 1)
+    # get decreasing order of ell axes in terms of dimension size
+    order = np.argsort([result.shape[i] for i in _axis])[::-1]
 
-    # get data & move ell axis to front
-    data = np.moveaxis(result, axis, 0)
+    # get axis in new order
+    axis = tuple(_axis[i] for i in order)
```

Comment:
why are the axis not in the correct order already?

Reply:
The first axis must have the longest `ell`, or we don't have enough rows to store the longest `ell` etc. vector.